### ISSUE TITLE
Update `ProtocolServer`'s exceptions to be more specific

### ISSUE THREAD

**Issue by NivekT** [pytorch/data#996]
Within `ProtocolServer` inside `dataloader2/communication/protocol.py`, the exceptions being raised inside are generic `Exception`. Ideally, we should change the exceptions to be more specific, such that they can be handled easily elsewhere.

cc: @ejguan

**Comment by ejguan** [comment]
Yeah. This is going to be important for us to differentiate the source of Error from `next(dp_iter)` or Error related to Request/Response
